MS Teams integration is unusable for new self-hosted installs: OAuth token authority hardcoded to botframework.com, but Microsoft deprecated multi-tenant bots (2025-07-31)

### Environment

self-hosted ([https://develop.sentry.dev/self-hosted/](<https://develop.sentry.dev/self-hosted/>))

### Steps to Reproduce

On a self-hosted Sentry (26.6.0), set up the Microsoft Teams integration from scratch, following [https://develop.sentry.dev/integrations/msteams/](<https://develop.sentry.dev/integrations/msteams/>)

1. Create a Teams app + bot in the Teams Developer Portal.
2. Create an **Azure Bot** resource for it (required — a Developer-Portal-only bot has no Bot Framework identity and can never send outbound messages). Azure today offers only `Single Tenant` or `User-Assigned Managed Identity` as "Type of App"; `Multi Tenant` **no longer exists**, since Microsoft deprecated multi-tenant bot creation on 2025-07-31.
   * Creation type: *Use existing app registration*
   * Backed by a **multi-tenant** Entra app registration ("Accounts in any organizational directory")
3. On the Azure Bot resource: set the messaging endpoint to `https://<sentry-host>/extensions/msteams/webhook/` and enable the **Microsoft Teams** channel.
4. Set `msteams.client-id` (bot app id) and `msteams.client-secret` in `config.yml`, restart `web`.
5. Publish the Teams app org-wide and add it to a team.

### Expected Result

Sentry posts the "Welcome to Sentry for Microsoft Teams" card to the channel, the webhook returns 200, and an `Integration` row with `provider='msteams'` is created.

### Actual Result

The inbound webhook is reached and passes JWT validation, but every **outbound** Bot Connector call returns **401**, so the webhook handler raises and returns **500**. The integration is never created and no welcome message is sent.

[INFO] sentry.access.api: api.access (method='POST' view='sentry.integrations.msteams.webhook.MsTeamsWebhookEndpoint' response='500' path='/extensions/msteams/webhook/' user_agent='Microsoft-SkypeBotApi (Microsoft-BotFramework/3.0)') [WARNING] sentry.integrations.client: integration.http_response (url='[https://smba.trafficmanager.net/at/](<https://smba.trafficmanager.net/at/>)/v3/conversations/19:...@thread.tacv2/activities' integration='msteams' status_string='401' error='401 Client Error: Unauthorized') [ERROR] django.request: Internal Server Error: /extensions/msteams/webhook/ (status_code=500)

### Root cause

Sentry acquires the Bot Connector token from the **multi-tenant** authority, hardcoded in  <br>[`src/sentry/integrations/msteams/client.py#L140`](<https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/msteams/client.py#L140>):

```python
class OAuthMsTeamsClient(ApiClient):
    TOKEN_URL = "/oauth2/v2.0/token"

    def __init__(self, client_id, client_secret):
        tenant_id = options.get("msteams.tenant-id")   # new option, defaults to ""
        self.base_url = (
            f"https://login.microsoftonline.com/{tenant_id}"
            if tenant_id
            else "https://login.microsoftonline.com/botframework.com"
        )
        ...
```

A single-tenant Azure Bot — now the only kind that can be created — must obtain its token from the tenant-specific authority instead:

[https://login.microsoftonline.com/](<https://login.microsoftonline.com/>)<TENANT_ID>/oauth2/v2.0/token  <br>There is no option to configure this, so the integration cannot work.

Proof (same credentials, same scope, same conversation — only the authority differs)

```python
SCOPE = "https://api.botframework.com/.default"
URL   = f"{service_url}/v3/conversations/{conversation_id}/members"   # read-only probe
```

# A) what Sentry does today

```python
tok_a = client_credentials("https://login.microsoftonline.com/botframework.com/oauth2/v2.0/token")
GET(URL, bearer=tok_a)   # -> 401 {"message":"Authorization has been denied for this request."}
```

# B) tenant-specific authority

```python
tok_b = client_credentials(f"https://login.microsoftonline.com/{TENANT_ID}/oauth2/v2.0/token")
GET(URL, bearer=tok_b)   # -> 200 OK
```

Both tokens carry aud=[https://api.botframework.com](<https://api.botframework.com>) and the correct appid; only the issuer differs.

Workarounds that do NOT exist  <br>Creating a multi-tenant Azure Bot: blocked by Microsoft since 2025-07-31.  <br>Flipping an existing bot to multi-tenant: az resource update --set properties.msaAppType=MultiTenant --remove properties.msaAppTenantId returns HTTP 200 with no error, but silently does not apply the change — msaAppType stays SingleTenant. Re-measuring afterwards confirms the 401 persists.  <br>Upgrading Sentry: master carries the same hardcoded line.  <br>Impact  <br>Any self-hosted instance configuring MS Teams for the first time after 2025-07-31 cannot complete the integration. Existing installs backed by legacy multi-tenant bots are unaffected, which likely explains why this has gone unnoticed. This may be the underlying cause of getsentry/sentry#108113.

Suggested fix (backwards compatible)  <br>Make the authority configurable; default to today's behaviour.

```python
class OAuthMsTeamsClient(ApiClient):
    TOKEN_URL = "/oauth2/v2.0/token"

    def __init__(self, client_id, client_secret):
        tenant_id = options.get("msteams.tenant-id")   # new option, defaults to ""
        self.base_url = (
            f"https://login.microsoftonline.com/{tenant_id}"
            if tenant_id
            else "https://login.microsoftonline.com/botframework.com"
        )
        ...

In src/sentry/options/defaults.py, next to the existing msteams.* options:
```python
register("msteams.tenant-id", flags=FLAG_PRIORITIZE_DISK | FLAG_AUTOMATOR_MODIFIABLE)
```

Existing multi-tenant deployments keep working untouched; single-tenant deployments set msteams.tenant-id in config.yml.

The docs at [https://develop.sentry.dev/integrations/msteams/](<https://develop.sentry.dev/integrations/msteams/>) should also state that an Azure Bot resource is mandatory — the current instructions stop at the Teams Developer Portal, which by itself produces a bot that can never send outbound messages.

References  <br>Multi-tenant bot deprecation (2025-07-31): [https://techcommunity.microsoft.com/discussions/teamsdeveloper/questions-about-alternatives-to-multi-tenant-bots-after-july-31-2025-deprecation/4423676  ](<https://techcommunity.microsoft.com/discussions/teamsdeveloper/questions-about-alternatives-to-multi-tenant-bots-after-july-31-2025-deprecation/4423676>)<br>Multi-tenant app + single-tenant bot → use tenant-specific authority: [https://learn.microsoft.com/en-us/answers/questions/5584294/multitenant-app-associated-to-single-tenant-azure  ](<https://learn.microsoft.com/en-us/answers/questions/5584294/multitenant-app-associated-to-single-tenant-azure>)<br>Developer-Portal-only bots cannot send outbound messages: [https://learn.microsoft.com/en-us/answers/questions/5650695/single-tenant-teams-bot-(dev-portal-only)-gets-401](<https://learn.microsoft.com/en-us/answers/questions/5650695/single-tenant-teams-bot-(dev-portal-only)-gets-401>)

### Product Area

Settings - Integrations

### Link

*No response*

### DSN

*No response*

### Version

26.6.0

--- a/src/sentry/integrations/msteams/client.py
+++ b/src/sentry/integrations/msteams/client.py
@@ -146,6 +146,12 @@ class OAuthMsTeamsClient(ApiClient):
         super().__init__()
         self.client_id = client_id
         self.client_secret = client_secret
+        # Single-tenant Azure Bots must request the token from their tenant's
+        # authority. Multi-tenant bots (the historical default) use the shared
+        # botframework.com authority, which Microsoft no longer provisions.
+        tenant_id = options.get("msteams.tenant-id")
+        if tenant_id:
+            self.base_url = f"https://login.microsoftonline.com/{tenant_id}"
 
     def exchange_token(self):
         headers = {"Content-Type": "application/x-www-form-urlencoded"}

--- a/src/sentry/integrations/msteams/client.py
+++ b/src/sentry/integrations/msteams/client.py
@@ -146,9 +146,8 @@ class OAuthMsTeamsClient(ApiClient):
         super().__init__()
         self.client_id = client_id
         self.client_secret = client_secret
-        # Single-tenant Azure Bots must request the token from their tenant's
-        # authority. Multi-tenant bots (the historical default) use the shared
-        # botframework.com authority, which Microsoft no longer provisions.
+        # tenant_id is required for single-tenant bots to authenticate. The bot framework
+        # base_url is specific to multi-tenant bots which microsoft no longer provisions.
         tenant_id = options.get("msteams.tenant-id")
         if tenant_id:
             self.base_url = f"https://login.microsoftonline.com/{tenant_id}"

--- a/src/sentry/options/defaults.py
+++ b/src/sentry/options/defaults.py
@@ -754,6 +754,9 @@ register("vercel.integration-slug", default="sentry", flags=FLAG_AUTOMATOR_MODIF
 register("msteams.client-id", flags=FLAG_PRIORITIZE_DISK | FLAG_AUTOMATOR_MODIFIABLE)
 register("msteams.client-secret", flags=FLAG_CREDENTIAL | FLAG_PRIORITIZE_DISK)
 register("msteams.app-id")
+# Tenant-specific OAuth authority, required for single-tenant Azure Bots.
+# Empty (default) keeps the historical multi-tenant botframework.com authority.
+register("msteams.tenant-id", flags=FLAG_PRIORITIZE_DISK | FLAG_AUTOMATOR_MODIFIABLE)
 
 # Discord Integration
 register("discord.application-id", flags=FLAG_PRIORITIZE_DISK | FLAG_AUTOMATOR_MODIFIABLE)

--- a/tests/sentry/integrations/msteams/test_client.py
+++ b/tests/sentry/integrations/msteams/test_client.py
@@ -9,7 +9,7 @@ from django.test import override_settings
 from requests import Request
 
 from sentry.integrations.models.integration import Integration
-from sentry.integrations.msteams.client import MsTeamsClient
+from sentry.integrations.msteams.client import MsTeamsClient, OAuthMsTeamsClient
 from sentry.shared_integrations.exceptions import IntegrationError
 from sentry.silo.base import SiloMode
 from sentry.silo.util import (
@@ -20,6 +20,7 @@ from sentry.silo.util import (
     PROXY_SIGNATURE_HEADER,
 )
 from sentry.testutils.cases import TestCase
+from sentry.testutils.helpers.options import override_options
 from sentry.testutils.silo import control_silo_test
 from tests.sentry.integrations.test_helpers import add_control_silo_proxy_response
 
@@ -287,3 +288,18 @@ class MsTeamsProxyApiClientTest(TestCase):
             assert request.headers[PROXY_BASE_URL_HEADER] == "https://smba.trafficmanager.net/amer"
             assert client.base_url not in request.url
             assert_proxy_request(request, is_proxy=True)
+
+
+@control_silo_test
+class OAuthMsTeamsClientTest(TestCase):
+    def test_defaults_to_botframework_authority(self) -> None:
+        client = OAuthMsTeamsClient("client-id", "client-secret")
+        assert client.base_url == "https://login.microsoftonline.com/botframework.com"
+
+    @override_options({"msteams.tenant-id": "00000000-1111-2222-3333-444444444444"})
+    def test_uses_tenant_authority_when_configured(self) -> None:
+        client = OAuthMsTeamsClient("client-id", "client-secret")
+        assert (
+            client.base_url
+            == "https://login.microsoftonline.com/00000000-1111-2222-3333-444444444444"
+        )

